pcl_ros/src/pcl_ros/filters/voxel_grid.cpp

```diff
@@ -35,100 +35,213 @@
-    NODELET_DEBUG("[config_callback] Setting the downsampling leaf size to: %f.", leaf_size[0]);
-    impl_.setLeafSize(leaf_size[0], leaf_size[1], leaf_size[2]);
-  }
+  unsigned int minPointsPerVoxel = impl_.getMinimumPointsNumberPerVoxel();
 
-  double filter_min, filter_max;
-  impl_.getFilterLimits(filter_min, filter_max);
-  if (filter_min != config.filter_limit_min) {
-    filter_min = config.filter_limit_min;
-    NODELET_DEBUG(
-      "[config_callback] Setting the minimum filtering value a point will be considered "
-      "from to: %f.",
-      filter_min);
+  for (const rclcpp::Parameter & param : params) {
+    if (param.get_name() == "filter_field_name") {
+      // Check the current value for the filter field
+      if (impl_.getFilterFieldName() != param.as_string()) {
+        // Set the filter field if different
+        impl_.setFilterFieldName(param.as_string());
+        RCLCPP_DEBUG(
+          get_logger(), "Setting the filter field name to: %s.",
+          param.as_string().c_str());
+      }
+    }
+    if (param.get_name() == "filter_limit_min") {
+      // Check the current values for filter min-max
+      if (filter_min != param.as_double()) {
+        filter_min = param.as_double();
+        RCLCPP_DEBUG(
+          get_logger(),
+          "Setting the minimum filtering value a point will be considered from to: %f.",
+          filter_min);
+        // Set the filter min-max if different
+        impl_.setFilterLimits(filter_min, filter_max);
+      }
+    }
+    if (param.get_name() == "filter_limit_max") {
+      // Check the current values for filter min-max
+      if (filter_max != param.as_double()) {
+        filter_max = param.as_double();
+        RCLCPP_DEBUG(
+          get_logger(),
+          "Setting the maximum filtering value a point will be considered from to: %f.",
+          filter_max);
+        // Set the filter min-max if different
+        impl_.setFilterLimits(filter_min, filter_max);
+      }
+    }
+    if (param.get_name() == "filter_limit_negative") {
+      bool new_filter_limits_negative = param.as_bool();
+      if (impl_.getFilterLimitsNegative() != new_filter_limits_negative) {
+        RCLCPP_DEBUG(
+          get_logger(),
+          "Setting the filter negative flag to: %s.",
+          (new_filter_limits_negative ? "true" : "false"));
+        impl_.setFilterLimitsNegative(new_filter_limits_negative);
+      }
+    }
+    if (param.get_name() == "min_points_per_voxel") {
+      if (minPointsPerVoxel != ((unsigned int) param.as_int())) {
+        RCLCPP_DEBUG(
+          get_logger(),
+          "Setting the minimum points per voxel to: %u.",
+          minPointsPerVoxel);
+        impl_.setMinimumPointsNumberPerVoxel(param.as_int());
+      }
+    }
+    // Defer decision on final leaf size until after all parameters are collected.
+    if (param.get_name() == "leaf_size") {
+      leaf_size.setConstant(param.as_double());
+    }
+    if (param.get_name() == "leaf_size_x") {
```

Comment:
This leaf size thing is very convoluted. You have this leaf_size array sized 3 you copy above. Why not just populate that here? You then check if `impl_.getLeafSize() != leaf_size` below, which would still trigger the same way without these additional 2nd level indirections

Reply:
I think we cannot assume the order of parameter iteration. If we have both `leaf_size` and `leaf_size_x`, the result can be nondeterministic.
How about using the original structure? Just putting the parameter values into `config_leaf_size`, `config_leaf_size_x`, ... in the loop and reuse the same logic here after the loop.

https://github.com/ros-perception/perception_pcl/blob/0d27ed78ee3a7071a6508516a166bab0640bcba8/pcl_ros/src/pcl_ros/filters/voxel_grid.cpp#L79-L81

Reply:
How about we remove this from this PR and handle separately since its a separate feature request this is adding

Reply:
Sure -- I can do that.